### PR TITLE
Flagged static (performance optimized) build

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Adds the new static build behind a flag `--experimental-static-build`.
- Does not affect normal build.
- The gist of how this works is:
  - The compiler lifts up dependencies statically, such as CSS which is not extracted separately from rendering.
  - The static build runs Vite build on each page `.astro` file. This gets us our CSS and imported images.
  - We `import()` the built astro page `dist/assets/index.mjs` for example.
  - We render it, providing the `link` stylesheets built.
- Note that this only works with the provided __examples/fast-build__ demo at the moment.
- Note that this does not build hydrated components or hoisted scripts at the moment.

## Testing

Example app

## Docs

N/A, this isn't close enough to being production ready to advertise.

### CODE REVIEW

File: examples/docs/src/layouts/MainLayout.astro [hunk 1]
```diff
@@ -118,5 +118,7 @@ const githubEditUrl = CONFIG.GITHUB_EDIT_URL && (CONFIG.GITHUB_EDIT_URL + curren
         <RightSidebar content={content} githubEditUrl={githubEditUrl} />
       </aside>
     </main>
+    <script type="module" src={Astro.resolve('./script.js')} hoist></script>
+    <img src={Astro.resolve('../images/twitter.png')} />
```
Comment (matthewp) [inline examples/docs/src/layouts/MainLayout.astro:122]: Note to self, remove these changes.

File: examples/docs/src/layouts/script.js [hunk 2]
```diff
@@ -0,0 +1 @@
+console.log("HERE I AM");
```
Comment (matthewp) [inline examples/docs/src/layouts/script.js:1]: remove this

File: packages/astro/test/astro-basic.test.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ describe('Astro basics', () => {
   });
 
   describe('build', () => {
-    it('Can load page', async () => {
+    it.only('Can load page', async () => {
```
Comment (matthewp) [inline packages/astro/test/astro-basic.test.js:21]: change this of course

**Review by jonathantneal — APPROVED** [review]
Many changes looked like outright rewrites, but the parts I could check the delta of were solid.

**Comment by FredKSchott** [comment]
**Looks great!** I ran this across the pokedex on a pretty lightweight page, and even still got a 50% speed reduction from 3sec to 1.5sec. I can't imagine how much better this would run on all of the large projects that are storing all of that data in memory.

**FYI:** I see an odd import path in the hydration script inside of `dist` when I run this, so there may be a bug in finally running the `dist` dir with `astro preview`:

```
const [{ default: Component }, { default: hydrate }] = await Promise.all([import("/Users/fks/Code/astro/examples/fast-build/dist/components/Greeting.vue"), import("@astrojs/renderer-vue/client.js")]);
```

**Lastly,** I'd love to see a `getStaticPaths()` example inside of `fast-build` so that we can quickly verify the perf as we go. You can use my example below:

```astro
---
// src/pages/[pokemon].astro
import Greeting from '../components/Greeting.vue';

export async function getStaticPaths() {
  const response = await fetch(`https://pokeapi.co/api/v2/pokemon?limit=150`);
  const result = await response.json();
  const allPokemon = result.results;
  return allPokemon.map(pokemon => ({params: {pokemon: pokemon.name}, props: {pokemon}}));
}
---
<html lang="en">
  <head>
    <title>Hello</title>
  </head>

  <body>
    <h1>{Astro.props.pokemon.name}</h1>
    <Greeting client:load />
  </body>
</html>
```

**Comment by matthewp** [comment]
@FredKSchott This PR does not implement hydrated scripts. That will come in a follow-up. Agree on adding getStaticPaths.

**Pull request merged**